# Исследование надежности заемщиков


<b>Описание проекта</b>

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

<b>Описание данных</b>

`children` — количество детей в семье

`days_employed` — общий трудовой стаж в днях

`dob_years` — возраст клиента в годах

`education` — уровень образования клиента

`education_id` — идентификатор уровня образования

`family_status` — семейное положение

`family_status_id` — идентификатор семейного положения

`gender` — пол клиента

`income_type` — тип занятости

`debt` — имел ли задолженность по возврату кредитов

`total_income` — ежемесячный доход

`purpose` — цель получения кредита

## Знакомство с данными

In [17]:
import pandas as pd

data = pd.read_csv('C:/datasets/2/data 2.csv')

In [18]:
# Вывожу первые 20 строчек датафрейма data на экран.
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

In [19]:
# Вывожу основную информацию о датафрейме
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

In [20]:
# Вывожу количество пропущенных значений для каждого столбца
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполню пропуски в этом столбце медианным значением по каждому типу из столбца `income_type`.

In [21]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

In [22]:
# Заменю все отрицательные значения положительными
data['days_employed'] = data['days_employed'].abs()

In [23]:
# Для каждого типа занятости выведу медианное значение трудового стажа в днях
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Однако этот столбец не понадобится вам для исследования.

In [24]:
# Выведу перечень уникальных значений столбца children
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

In [25]:
# В столбце children есть два аномальных значения. Удалю эти строки из датафрейма.
data = data[(data['children'] != -1) & (data['children'] != 20)]

In [26]:
# Вывожу перечень уникальных значений столбца children, чтобы убедиться, что артефакты удалены
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

In [27]:
# Заполню пропуски в столбце days_employed медианными значениями по каждого типа занятости income_type.
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

Чтобы убедиться, что все пропуски заполнены. Ещё раз выведу количество пропущенных значений для каждого столбца.

In [28]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

In [29]:
# Заменяю вещественный тип данных в столбце total_income на целочисленный
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

Обработаю неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведу их к нижнему регистру и проверю остальные столбцы.

In [30]:
# Привожу значения в столбце к нижнему регистру
data['education'] = data['education'].str.lower()

In [31]:
# Вывожу на экран количество строк-дубликатов в данных
data.duplicated().sum()

71

In [32]:
# Удаляю дубликаты
data = data.drop_duplicates()

### Категоризация данных

Создам в датафрейме `data` столбец `total_income_category` с категориями на основании диапазонов, указанных ниже:

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


In [33]:
# Создаю функцию для категоризации
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [34]:
# Применяю функцию к столбцу
data['total_income_category'] = data['total_income'].apply(categorize_income)

In [35]:
# Вывожу на экран перечень уникальных целей взятия кредита из столбца purpose
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

Создам функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

In [36]:
# Создаю функцию для категоризации
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [37]:
# Применяю функцию к столбцу
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Исследование данных

**Необходимо понять, есть ли зависимость между количеством детей и возвратом кредита в срок.**

In [38]:
# Группирую данные по количеству детей и нахожу среднее значение задолженности для каждого случая.

print(data.groupby('children').agg({'debt':'mean'}))

              debt
children          
0         0.075438
1         0.092346
2         0.094542
3         0.081818
4         0.097561
5         0.000000


**Вывод:** 

Видно, что отсутствие детей позитивно влияет на уплату долгов.

Однако семьи с тремя детьми по какой-то причине имеют меньше долгов, чем семьи с одни ребенком или двумя детьми. Скорее всего, таких семей попросту меньше, поэтому и выборка меньше. Похожая аномалия наблюдается и в отношении семей с 5 детьми.

Другим объяснением может быть то, что для обычных семей, рождение ребенка — это некий шок, влияющий на исполнение финансовых обязательств. В то время как для семей с большим количеством детей дети стали обыденностью, поэтому они научились распределять семейный бюджет более качественно

**Есть ли зависимость между семейным положением и возвратом кредита в срок.**

In [39]:
# Группирую данные по семейному положению и нахожу среднее значение задолженности для каждого случая.

print(data.groupby('family_status').agg({'debt':'mean'}).sort_values(by='debt', ascending=True))

                           debt
family_status                  
вдовец / вдова         0.066246
в разводе              0.070648
женат / замужем        0.075606
гражданский брак       0.093130
Не женат / не замужем  0.097639


**Вывод:** 

Люди, никогда не вступавшие в брачные отношения, хуже возвращают долги.

Кредитному отделу следует с осторожностью относиться к клиентам, которые не состоят или не состояли в браке. Согласно данным такие клиенты менее ответственные. Получается, что узаконенные отношения ведут к более ответственному отношению к своим обязательствам. 

**Существует ли зависимость между уровнем дохода и возвратом кредита в срок?**

In [40]:
# Группирую данные по категориям дохода и нахожу среднее значение задолженности для каждой категории.

print(data.groupby('total_income_category').agg({'debt':'mean'}).sort_values(by='debt', ascending=True))

                           debt
total_income_category          
D                      0.060172
B                      0.070602
A                      0.080000
C                      0.084982
E                      0.090909


**Вывод:** 

При уровне дохода от 30 001 до 50 000 человек чаще всего выплачивает свои долги, однако, если его доходы ниже категории D, он с наименьшей вероятностью вернет долг. При этом повышение доходов выше категории D тоже негативно сказывается на уплате долгов.

Исходя из здравого смысла логично предположить, что чаще всего должниками становятся люди с небольшим доходом, которым финансовые проблемы не позволяют вовремя осуществлять платежи. Однако выборка по заемщикам, как с наименьшим доходом, так и с наибольшим крайне недостаточна для формирования однозначных выводов. Можно думать, что E-категория самая низкооплачиваемая не может корректно распоряжаться деньгами и «влезает» в кредиты, но тогда неясно почему у А-категории также достаточно высокий процент задолженности. Таким образом, по группам А и Е вывод неочевиден, поскольку недостаточно данных.

Если же взять для анализа только группы В, С и D, то по полученным данным не наблюдаем зависимости "те, кто получает больше ЗП, лучше возвращают кредит". В целом это логично, так как материальный достаток - не единственная переменная, которая влияет на качество кредитной истории.

Однако 2174 пропуска в данных столбца "total_income" ставят под сомнение полученный нами результат. Сводная таблица (или группировка) с данными о зависимости между уровнем дохода и возвратом кредита в срок возможно могла бы выглядеть иначе, если бы не пропущенные данные.

**Как разные цели кредита влияют на его возврат в срок?**

In [41]:
# Группирую данные по целям взятия кредита и нахожу среднее значение задолженности для каждой цели.

print(data.groupby('purpose_category').agg({'debt':'mean'}).sort_values(by='debt', ascending=True))


                              debt
purpose_category                  
операции с недвижимостью  0.072551
проведение свадьбы        0.079118
получение образования     0.092528
операции с автомобилем    0.093480


**Вывод:** 

Активнее всего клиенты возвращают долги за операции с недвижимостью и проведение свадьбы.

Клиенты реже возвращают долги по кредитам, связанным с получением образования. Возможно, потому, что получение образования - длительный процесс и до его завершения сложно устроиться на высокооплачиваемую работу.

Клиенты наиболее редко возвращают долги по кредитам, связанным с операциями с автомобилем. Возможно, потому, что содержание автомобиля является дополнительной статьёй расходов.

#### Причины появления пропусков в исходных данных.

Пропуски располагались в столбцах `days_employed` (стаж) и `total_income` (доход), причем количество пропусков в данных столбцах одинаково, пропуски в строках идут попарно, то есть с пропуском в столбце `days_employed` всегда соседствует пропуск в `total_income`. 

1) Возможно, это связано с тем, что данные клиенты сменили работу и не предоставили банку новой информации. 

2) Возможно, произошла техническая ошибка при записи или копировании данных, и информация о стаже и доходе данных клиентов пропала.

#### Почему заполнить пропуски медианным значением — лучшее решение для количественных переменных?

Потому что это наиболее верно передаёт реальное положение вещей. Среднее значение, в отличие от медианного, может быть сильно завышено или заниженно, если в данных есть аномально низкие или высокие позиции.


### Общий вывод.

Было проведено исследование, цель которого - поиск ответов на следующие вопросы:
    1)Есть ли зависимость между количеством детей и возвратом кредита в срок?
    2)Есть ли зависимость между семейным положением и возвратом кредита в срок?
    3)Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
    4)Как разные цели кредита влияют на его возврат в срок?

<b>1. Зависимость между количеством детей и возвратом кредита в срок существует</b>
Отсутствие детей позитивно влияет на уплату долгов.

    7,5% - доля задолженностей у клиентов без детей.
    9,2% - доля задолженностей у клиентов с одним ребенком.
    9,4% - доля задолженностей у клиентов с двумя детьми.

Рекомендация: для получения информации о клиентах с 3,4 и 5 детьми необходимо больше данных.

<b>2. Зависимость между семейным положением и возвратом кредита в срок существует</b>
Люди, никогда не вступавшие в брачные отношения, хуже возвращают долги.

    6,6% - доля задолженностей у овдовевших клиентов.
    7% - доля задолженностей у разведённых клиентов.
    7,5% - доля задолженностей у женатых/замужних клиентов.
    9,3% - доля задолженностей у клиентов, состоящих в гражданском браке.
    9,7% - - доля задолженностей у неженатых/незамужних клиентов.

<b>3. Зависимость между уровнем дохода и возвратом кредита в срок существует</b>
При уровне дохода от 30 001 до 50 000 человек чаще всего выплачивает свои долги, однако, если доходы клиента ниже 30 001, он с наименьшей вероятностью вернёт долг. 

    6% - доля задолженностей у клиентов с доходом от 30 001 до 50 000 р.
    9% - доля задолженностей у клиентов с доходом до 30 000 р.

Рекомендация: для получения информации о клиентах с более высоким доходом необходимо исследовать больше данных.

<b>4. Разные цели кредита влияют на его возврат в срок.</b>
Активнее всего клиенты возвращают долги за операции с недвижимостью и проведение свадьбы. Реже возвращают долги по кредитам, связанным с получением образования. Наиболее редко возвращают долги по кредитам, связанным с операциями с автомобилем. 

    7,2% - доля задолженностей у клиентов, взявших кредит на операции с недвижимостью.  
    7,9% - доля задолженностей у клиентов, взявших кредит на проведение свадьбы.
    9,2% - доля задолженностей у клиентов, взявших кредит на получение образования.
    9,3% - доля задолженностей у клиентов, взявших кредит на операции с автомобилем.

<b>Общие рекомендации:</b>
Специалистам, которые занимаются построением системы кредитного скоринга, будут полезны выводы, полученные о целях кредита (вопрос №4), о семейном положении (вопрос №2), а также отсутствии детей у клиентов (вопрос №1).

Для формулирования более чётких и полных выводов насчёт связи количества детей (вопрос №1) и дохода клиентов (вопрос №3) с их задолженностями необходимо больше данных.
